crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,542 @@
+    }
+
+    pub fn lock(&self) {
+        unsafe { pw_thread_loop_lock(self.0.as_ptr()) }
+    }
+
+    pub fn start(&self) {
+        unsafe {
+            pw_thread_loop_start(self.0.as_ptr());
+        }
+    }
+
+    pub fn signal(&self) {
+        unsafe {
+            pw_thread_loop_signal(self.0.as_ptr(), false);
+        }
+    }
+
+    pub fn wait(&self) {
+        unsafe {
+            pw_thread_loop_wait(self.0.as_ptr());
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PwGetLoop(NonNull<pw_loop>);
+
+impl AsRef<LoopRef> for PwGetLoop {
+    fn as_ref(&self) -> &LoopRef {
+        self.deref()
+    }
+}
+
+impl Deref for PwGetLoop {
+    type Target = LoopRef;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(self.0.as_ptr() as *mut LoopRef) }
+    }
+}
+
+// SAFETY: Safe as the structure can be sent to another thread.
+unsafe impl Send for PwBackend {}
+
+// SAFETY: Safe as the structure can be shared with another thread as the state
+// is protected with a lock.
+unsafe impl Sync for PwBackend {}
 
 pub struct PwBackend {
-    streams: Arc<RwLock<Vec<Stream>>>,
+    pub stream_params: Arc<RwLock<Vec<Stream>>>,
+    thread_loop: Arc<PwThreadLoop>,
+    pub core: Core,
+    #[allow(dead_code)]
+    context: Context<PwGetLoop>,
+    pub stream_hash: RwLock<HashMap<u32, pw::stream::Stream>>,
+    pub stream_listener: RwLock<HashMap<u32, pw::stream::StreamListener<i32>>>,
 }
 
 impl PwBackend {
-    pub fn new(streams: Arc<RwLock<Vec<Stream>>>) -> Self {
-        Self { streams }
+    pub fn new(stream_params: Arc<RwLock<Vec<Stream>>>) -> Self {
+        pw::init();
+
+        let thread_loop = Arc::new(PwThreadLoop::new(Some("Pipewire thread loop")).unwrap());
+        let get_loop = thread_loop.get_loop();
+
+        thread_loop.lock();
+
+        let context = pw::Context::new(&get_loop).expect("failed to create context");
+        thread_loop.start();
+        let core = context.connect(None).expect("Failed to connect to core");
+
+        // Create new reference for the variable so that it can be moved into the closure.
+        let thread_clone = thread_loop.clone();
+
+        // Trigger the sync event. The server's answer won't be processed until we start the thread loop,
+        // so we can safely do this before setting up a callback. This lets us avoid using a Cell.
+        let pending = core.sync(0).expect("sync failed");
+        let _listener_core = core
+            .add_listener_local()
+            .done(move |id, seq| {
+                if id == PW_ID_CORE && seq == pending {
+                    thread_clone.signal();
+                }
+            })
+            .register();
+
+        thread_loop.wait();
+        thread_loop.unlock();
+
+        log::trace!("pipewire backend running");
+
+        Self {
+            stream_params,
+            thread_loop,
+            core,
+            context,
+            stream_hash: RwLock::new(HashMap::with_capacity(NR_STREAMS)),
```

Comment:
I think the number of stream can be got from `stream_params.len()` thus NR_STREAMS can be removed.